scanner/src/main/kotlin/experimental/WorkingTreeCache.kt

```diff
@@ -0,0 +1,93 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.experimental
+
+import java.io.IOException
+
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+import org.ossreviewtoolkit.downloader.VersionControlSystem
+import org.ossreviewtoolkit.downloader.WorkingTree
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.utils.common.safeDeleteRecursively
+import org.ossreviewtoolkit.utils.core.createOrtTempDir
+
+/**
+ * A cache for [VCS][VersionControlSystem] [WorkingTree]s that manages one working tree for each triple of
+ * [VcsInfo.type], [VcsInfo.url], and [VcsInfo.path].
+ */
+interface WorkingTreeCache {
+    /**
+     * Apply the [block] to the working tree defined by [vcsInfo]. The [VcsInfo.revision] is ignored during
+     * initialization of the working tree, it is the caller's responsibility to update the working tree to the correct
+     * revision. It is ensured that only one [block] at a time can access a single working tree, separate working trees
+     * can be accessed in parallel.
+     */
+    suspend fun <T> use(vcsInfo: VcsInfo, block: (VersionControlSystem, WorkingTree) -> T): T
+
+    /**
+     * Clear all cached working trees from the file system. The function waits for all currently running operations on
+     * the working trees to finish before deleting them.
+     */
+    suspend fun clear()
+}
+
+class DefaultWorkingTreeCache : WorkingTreeCache {
+    private val mutex = Mutex()
+    private val workingTreeMutexes = mutableMapOf<String, Mutex>()
+    private val workingTrees = mutableMapOf<String, WorkingTree>()
+
+    override suspend fun <T> use(vcsInfo: VcsInfo, block: (VersionControlSystem, WorkingTree) -> T): T =
+        getWorkingTreeMutex(vcsInfo).withLock {
+            val vcs = getVcs(vcsInfo)
+            val workingTree = getWorkingTree(vcsInfo)
+
+            block(vcs, workingTree)
+        }
+
+    private fun getKey(vcsInfo: VcsInfo) = "${vcsInfo.type}|${vcsInfo.url}|${vcsInfo.path}"
+
+    private suspend fun getWorkingTreeMutex(vcsInfo: VcsInfo) =
+        mutex.withLock { workingTreeMutexes.getOrPut(getKey(vcsInfo)) { Mutex() } }
+
+    private fun getVcs(vcsInfo: VcsInfo) =
+        VersionControlSystem.forType(vcsInfo.type)
+            ?: VersionControlSystem.forUrl(vcsInfo.url)
+            ?: throw IOException("Could not determine VCS type for (type=$${vcsInfo.type}, url=${vcsInfo.url}).")
+
+    private fun getWorkingTree(vcsInfo: VcsInfo): WorkingTree {
+        val vcs = getVcs(vcsInfo)
+        val dir = createOrtTempDir()
+        return workingTrees.getOrPut(getKey(vcsInfo)) { vcs.initWorkingTree(dir, vcsInfo.copy(revision = "")) }
+    }
+
+    override suspend fun clear() {
```

Comment:
Would it make sense to turn `clear()` into `shutdown()`, in particular not allow any `use` call after `clear()` returns?

This way the caller of `clear()` can be sure that storage space is actually cleared as soon as his code execution continues after returning from this blocking call.

Reply:
It's another approach, but making `clear()` non-blocking also has benefits, it allows to continue operation while the data is deleted in the background (even though currently it's only called inside `runBlocking` anyway).

Reply:
What do you mean by "making `clear()` non-locking"?

To my understanding it is blocking because it acquires locks.

Reply:
It's still a suspend function, the mutex lock only blocks other coroutines from entering the same code block, but they can continue doing other stuff. See: https://kotlinlang.org/docs/shared-mutable-state-and-concurrency.html#mutual-exclusion

Reply:
Currently this does not make a difference because all provenance resolution is done in order, but I want to change this to use coroutines later on for better performance.

Reply:
I like the idea to prevent any `use()` calls after an invocation of `clear()`. Otherwise, there is always a potential for a race condition that a `use` is processed after a `clean` leaving some data on the disk. This could be achieved by setting a flag in `clear()`. `use()` would check this flag and fail it if was *true*.

Reply:
Ok, as both of you prefer that approach I have changed it.